pkg/compactor/blocks_cleaner.go

```diff
@@ -391,3 +408,49 @@ func (c *BlocksCleaner) cleanUserPartialBlocks(ctx context.Context, partials map
 		level.Info(userLogger).Log("msg", "deleted partial block marked for deletion", "block", blockID)
 	}
 }
+
+// applyUserRetentionPeriod marks blocks for deletion which have aged past the retention period.
+func (c *BlocksCleaner) applyUserRetentionPeriod(ctx context.Context, idx *bucketindex.Index, userID string, userBucket *bucket.UserBucketClient, userLogger log.Logger) {
+	retention := c.cfgProvider.CompactorBlocksRetentionPeriod(userID)
+
+	// The retention period of zero is a special value indicating to never delete.
+	if retention <= 0 {
+		return
+	}
+
+	level.Debug(userLogger).Log("msg", "applying retention", "retention", retention.String())
+	blocks := listBlocksOutsideRetentionPeriod(idx, time.Now(), retention)
+
+	// Attempt to mark all blocks. It is not critical if a marking fails, as
+	// the cleaner will retry applying the retention in its next cycle.
+	for _, b := range blocks {
+		level.Info(userLogger).Log("msg", "applied retention: marking block for deletion", "block", b.ID, "maxTime", b.MaxTime)
+		if err := block.MarkForDeletion(ctx, userLogger, userBucket, b.ID, fmt.Sprintf("block exceeding retention of %v", retention), c.blocksMarkedForDeletion); err != nil {
+			level.Warn(userLogger).Log("msg", "failed to mark block for deletion", "block", b.ID, "err", err)
+		}
+	}
+}
+
+// listBlocksOutsideRetentionPeriod determines the blocks which have aged past
+// the specified retention period, and are not already marked for deletion.
+func listBlocksOutsideRetentionPeriod(idx *bucketindex.Index, now time.Time, retention time.Duration) (result bucketindex.Blocks) {
```

Comment:
Nit: There is no need to pass both `now` and `retention` to this function. The only value that we really need here is "now - retention".

Reply:
The only reason for doing this was so that the unit test for the function had coverage of the subtraction, but I'm happy to change it as I think @pracucci has the same comment.